bookkeeper-server/src/main/java/org/apache/bookkeeper/proto/checksum/DigestManager.java

```diff
@@ -99,41 +98,22 @@ public static DigestManager instantiate(long ledgerId, byte[] passwd, DigestType
      */
     public ByteBufList computeDigestAndPackageForSending(long entryId, long lastAddConfirmed, long length,
             ByteBuf data) {
+        ByteBuf headersBuffer;
         if (this.useV2Protocol) {
-            /*
-             * For V2 protocol, use pooled direct ByteBuf's to avoid object allocation in DigestManager.
-             */
-            ByteBuf headersBuffer = allocator.buffer(METADATA_LENGTH + macCodeLength);
-            headersBuffer.writeLong(ledgerId);
-            headersBuffer.writeLong(entryId);
-            headersBuffer.writeLong(lastAddConfirmed);
-            headersBuffer.writeLong(length);
-
-            update(headersBuffer);
-            update(data);
-            populateValueAndReset(headersBuffer);
-
-            return ByteBufList.get(headersBuffer, data);
+            headersBuffer = allocator.buffer(METADATA_LENGTH + macCodeLength);
         } else {
-            /*
-             * For V3 protocol, use unpooled heap ByteBuf's (backed by accessible array): The one object
-             * allocation here saves us later allocations when converting to protobuf ByteString.
-             */
-            ByteBuf sendBuffer = Unpooled.buffer(METADATA_LENGTH + macCodeLength + data.readableBytes());
-            sendBuffer.writeLong(ledgerId);
-            sendBuffer.writeLong(entryId);
-            sendBuffer.writeLong(lastAddConfirmed);
-            sendBuffer.writeLong(length);
-
-            update(sendBuffer);
-            update(data);
-            populateValueAndReset(sendBuffer);
-
-            sendBuffer.writeBytes(data, data.readerIndex(), data.readableBytes());
-            ReferenceCountUtil.release(data);
-
-            return ByteBufList.get(sendBuffer);
+            headersBuffer = Unpooled.buffer(METADATA_LENGTH + macCodeLength);
```

Comment:
Should this be
`ByteBuf sendBuffer = Unpooled.buffer(METADATA_LENGTH + macCodeLength + data.readableBytes());` instead?

Reply:
no, whole point of this change is not to allocate new byte[] to concatenate header and data